### PR TITLE
Better separation of legacy and default token cache

### PR DESCRIPTION
**Pull request by oldalton**
Still fixing tests and need to resolve conflicts, but putting it out there if someone wants to review :)

### CODE REVIEW

File: IdentityCore/src/cache/key/MSIDLegacyTokenCacheKey.m [hunk 1]
```diff
@@ -38,12 +38,12 @@ @implementation MSIDLegacyTokenCacheKey
 #pragma mark - Helpers
 
 //We should not put nil keys in the keychain. The method substitutes nil with a special GUID:
-+ (NSString *)getAttributeName:(NSString *)original
+- (NSString *)getAttributeName:(NSString *)original
```
Comment (unpluggedk) [inline IdentityCore/src/cache/key/MSIDLegacyTokenCacheKey.m]: original? what is so original about thig? :p

File: IdentityCore/src/cache/token/MSIDAccountCacheItem.m [hunk 2]
```diff
@@ -22,95 +22,143 @@
...
-    // Account type
+
+    _json = json;
+
     _accountType = [MSIDAccountTypeHelpers accountTypeFromString:json[MSID_AUTHORITY_TYPE_CACHE_KEY]];
-    
-    // Extensibility
-    _additionalAccountFields = json;
-    
+    _legacyUserId = json[MSID_ACCOUNT_ID_CACHE_KEY];
```
Comment (unpluggedk) [inline IdentityCore/src/cache/token/MSIDAccountCacheItem.m]: General Q not particularly specific to this.
If some constants are only used in one class, do they really need to be classified constants outside?

File: IdentityCore/src/oauth2/token/MSIDCredentialType.h [hunk 3]
```diff
@@ -22,20 +22,20 @@
 // THE SOFTWARE.
 
 #import <Foundation/Foundation.h>
-#import "MSIDSharedCacheAccessor.h"
-#import "MSIDBaseToken.h"
 
-@interface MSIDTestCacheAccessor : NSObject <MSIDSharedCacheAccessor>
+typedef NS_ENUM(NSInteger, MSIDCredentialType)
+{
+    MSIDCredentialTypeOther = 0,
```
Comment (unpluggedk) [inline IdentityCore/src/oauth2/token/MSIDCredentialType.h]: What is included in Other?

File: IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m [hunk 4]
```diff
@@ -22,59 +22,350 @@
...
+    return [self saveSSOStateWithFactory:factory requestParams:requestParams response:response context:context error:error];
+}
+
+- (BOOL)saveTokensWithFactory:(MSIDOauth2Factory *)factory
+               brokerResponse:(MSIDBrokerResponse *)response
+             saveSSOStateOnly:(BOOL)saveSSOStateOnly
+                      context:(id<MSIDRequestContext>)context
+                        error:(NSError **)error
+{
+    // MSAL currently doesn't yet support broker
```
[the remark this answers is not in the document]
Comment (oldalton) [reply IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m]: good idea, added

File: IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m [hunk 5]
```diff
@@ -95,768 +386,510 @@ - (void)fillInternalErrorWithMessage:(NSString *)message
...
-        return NO;
+        return [self saveAccount:account context:context error:error];
     }
-    
-    return [self saveTokenWithPreferredCache:refreshToken
-                                     account:account
-                                     context:context
-                                       error:error];
+
+    return YES;
```
[the remark this answers is not in the document]
Comment (oldalton) [reply IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m]: No

File: IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m [hunk 6]
```diff
@@ -95,768 +386,510 @@ - (void)fillInternalErrorWithMessage:(NSString *)message
...
-    MSID_LOG_INFO(context, @"(Default accessor) Saving account");
-    MSID_LOG_INFO_PII(context, @"(Default accessor) Saving account %@", account);
-    
-    // Save account
-    return [self saveAccount:account
-               requestParams:requestParams
-                     context:context
-                       error:error];
+
+    return YES;
```
[the remark this answers is not in the document]
Comment (oldalton) [reply IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m]: No

File: IdentityCore/src/cache/key/MSIDDefaultCredentialCacheQuery.m [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+- (NSString *)service
+{
+    if (self.matchAnyCredentialType)
+    {
+        return nil;
+    }
+
+    switch (self.credentialType)
+    {
+        case MSIDCredentialTypeAccessToken:
```
[the remark this answers is not in the document]
Comment (oldalton) [reply IdentityCore/src/cache/key/MSIDDefaultCredentialCacheQuery.m]: renamed

File: IdentityCore/src/oauth2/token/MSIDCredentialType.h (hunk 3 above)
Comment (oldalton) [reply IdentityCore/src/oauth2/token/MSIDCredentialType.h]: Anything else we find in cache and can't translate to anything else :) A "default" value, if we want to extend more types in future

File: IdentityCore/tests/util/cache/MSIDTestTokenResponse.m [hunk 8]
```diff
@@ -232,10 +254,28 @@ + (MSIDTokenResponse *)defaultTokenResponseWithAT:(NSString *)accessToken
...
     NSString *scopesString = scopes.msidToString;
+
+    NSMutableDictionary *dictionary = [@{@"token_type": @"Bearer",
+                                         @"expires_in": @3600,
+                                         @"scope": scopesString
+                                         } mutableCopy];
+
+    if (accessToken)
+    {
+        dictionary[@"access_token"] = accessToken;
```
Comment (unpluggedk) [inline IdentityCore/tests/util/cache/MSIDTestTokenResponse.m]: nit: do we want to define the string values used in this method?

**Pull request merged**